keps/sig-architecture/20191023-conformance-without-beta.md

```diff
@@ -0,0 +1,163 @@
+- [Proposal](#proposal)
+  - [User Stories](#user-stories)
+    - [Story 1](#story-1)
+    - [Story 2](#story-2)
+    - [Story 3](#story-3)
+- [Design Details](#design-details)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Infrastructure Needed](#infrastructure-needed)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+**ACTION REQUIRED:** In order to merge code into a release, there must be an issue in [kubernetes/enhancements] referencing this KEP and targeting a release milestone **before [Enhancement Freeze](https://github.com/kubernetes/sig-release/tree/master/releases)
+of the targeted release**.
+
+For enhancements that make changes to code or processes/procedures in core Kubernetes i.e., [kubernetes/kubernetes], we require the following Release Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These checklist items _must_ be updated for the enhancement to be released.
+
+- [x] kubernetes/enhancements issue in release milestone, which links to KEP (this should be a link to the KEP location in kubernetes/enhancements, not the initial KEP PR)
+- [ ] KEP approvers have set the KEP status to `implementable`
+- [ ] Design details are appropriately documented
+- [ ] Test plan is in place, giving consideration to SIG Architecture and SIG Testing input
+- [ ] Graduation criteria is in place
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+The KEP describes a process for ensuring that the Kubernetes components and the Kubernetes conformance tests have no dependencies on beta REST APIs or features.
+
+## Motivation
+
+As the Kubernetes project matures, and is used as the foundation for other projects,
+production distributions are expected to increase in stability, reliability, and consistency.
+The Kubernetes conformance project is an important aspect of ensuring consistency across distributions.
+A production distribution should not be required to enable non-GA features or REST APIs in order to pass conformance tests.
+
+Beta dependencies can be grouped into several categories:
+* Kubernetes components requiring beta REST APIs with no stable alternatives (for example, kubelets use the beta CertificateSigningRequest endpoints for cert rotation)
```

Comment:
FWIW we've been blocked on apimachinery folks for this and bootstrap tokens forever now.  
cc @neolit123 @fabriziopandini

Reply:
CertificateSigningRequest I'm aware of. What is blocking on bootstrap tokens?

Reply:
Well assign you on the issues.

Reply:
i believe @timothysc and @luxas originally had the idea to make the BoostrapToken(String) objects available outside of kubeadm.

https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/types.go#L247-L269

i'm not aware what prior discussions happened on this topic.

this is related to:
https://github.com/kubernetes/kubernetes/pull/77669
and
https://github.com/kubernetes/kubeadm/issues/964

boostrap tokens being essentially a type of Secret (bootstrap.kubernetes.io/token) puts them in a bit of weird spot.

if the utilities here:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/apis/kubeadm/bootstraptokenhelpers.go
don't get exposed outside of kubeadm, then it feels like a "boostrap token" will be an abstract type, due to the mostly undocumented "stringData" usage?

yet, we also have the option to:
- close https://github.com/kubernetes/kubeadm/issues/964
- close https://github.com/kubernetes/kubernetes/pull/77669
- claim that the abstract type "bootstrap token" is GA because kubeadm, as the only publicly known consumer is GA and v1.Secret is GA.
- document "stringData" for tokens
- leave the token utils inside the kubeadm api under a seperate folder (cc @rosti, @dixudx )

Reply:
thanks, that's helpful context. let's take this discussion to https://github.com/kubernetes/kubernetes/pull/77669